src/crates/registry.rs

```diff
@@ -0,0 +1,216 @@
+
+static CRATES_ROOT: &str = "https://static.crates.io/crates";
+
+impl RegistryCrate {
+    pub(super) fn new(registry: Registry, name: &str, version: &str) -> Self {
+        RegistryCrate {
+            registry,
+            name: name.into(),
+            version: version.into(),
+        }
+    }
+
+    fn cache_path(&self, workspace: &Workspace) -> PathBuf {
+        workspace
+            .cache_dir()
+            .join(self.registry.cache_folder())
+            .join(&self.name)
+            .join(format!("{}-{}.crate", self.name, self.version))
+    }
+
+    fn fetch_url(&self, workspace: &Workspace) -> Result<String, Error> {
+        match &self.registry {
+            Registry::CratesIo => Ok(format!(
+                "{0}/{1}/{1}-{2}.crate",
+                CRATES_ROOT, self.name, self.version
+            )),
+            Registry::Alternative(alt) => {
+                let index_path = workspace
+                    .cache_dir()
+                    .join("registry-index")
+                    .join(alt.index_folder());
+                if !index_path.exists() {
+                    let url = alt.index();
+                    git2::Repository::clone(url, index_path.clone())
+                        .with_context(|_| format!("unable to update_index at {}", url))?;
+                    info!("cloned registry index");
+                }
+                let config = std::fs::read_to_string(index_path.join("config.json"))?;
+                let template_url = serde_json::from_str::<IndexConfig>(&config)
+                    .context("registry has invalid config.json")?
+                    .dl;
+                let replacements = [("{crate}", &self.name), ("{version}", &self.version)];
+
+                let url = if replacements
+                    .iter()
+                    .any(|(key, _)| template_url.contains(key))
+                {
+                    let mut url = template_url;
+                    for (key, value) in &replacements {
+                        url = url.replace(key, value);
+                    }
+                    url
+                } else {
+                    format!("{}/{}/{}/download", template_url, self.name, self.version)
+                };
+
+                Ok(url)
+            }
+        }
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct IndexConfig {
+    dl: String,
+}
+
+pub struct AlternativeRegistry {
+    registry_index: String,
+}
+
+impl AlternativeRegistry {
+    pub fn new(registry_index: impl Into<String>) -> AlternativeRegistry {
+        AlternativeRegistry {
+            registry_index: registry_index.into(),
+        }
+    }
+
+    fn index(&self) -> &str {
+        self.registry_index.as_str()
+    }
+
+    fn index_folder(&self) -> String {
+        // https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits
+        self.registry_index.as_str().replace(
+            &['/', '\\', '<', '>', '|', '?', '*', '"', ':', '.'][..],
+            &"-",
+        )
```

Comment:
What we did in other parts of rustwide is using the `percent_encoding` crate to urlencode the paths. [See an example in the git repos support.](https://github.com/rust-lang/rustwide/blob/d2c0723c88a7b05715b4942226acebf54d0b6328/src/crates/git.rs#L53-L58) What I'd do is move the definition of `ENCODE_SET` out of that module and into `crate::utils`, and use the same mechanism here.